plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/OracleClientModule.java

```diff
@@ -36,36 +37,67 @@
-        bindSessionPropertiesProvider(binder, OracleSessionProperties.class);
-        configBinder(binder).bindConfig(OracleConfig.class);
-        newOptionalBinder(binder, Key.get(int.class, MaxDomainCompactionThreshold.class)).setBinding().toInstance(ORACLE_MAX_LIST_EXPRESSIONS);
-        newSetBinder(binder, ConnectorTableFunction.class).addBinding().toProvider(Query.class).in(Scopes.SINGLETON);
+        bind(JdbcClient.class).annotatedWith(ForBaseJdbc.class).to(OracleClient.class).in(SINGLETON);
+        bindSessionPropertiesProvider(binder(), OracleSessionProperties.class);
+        configBinder(binder()).bindConfig(OracleConfig.class);
+        newOptionalBinder(binder(), Key.get(int.class, MaxDomainCompactionThreshold.class)).setBinding().toInstance(ORACLE_MAX_LIST_EXPRESSIONS);
+        newSetBinder(binder(), ConnectorTableFunction.class).addBinding().toProvider(Query.class).in(SINGLETON);
+        install(new RetryingConnectionFactoryModule());
+        install(new HideDetailsModule());
     }
 
-    @Provides
-    @Singleton
-    @ForBaseJdbc
-    public static ConnectionFactory connectionFactory(BaseJdbcConfig config, CredentialProvider credentialProvider, OracleConfig oracleConfig, OpenTelemetry openTelemetry)
-            throws SQLException
+    private static class HideDetailsModule
+            extends PrivateModule
     {
-        Properties connectionProperties = new Properties();
-        connectionProperties.setProperty(OracleConnection.CONNECTION_PROPERTY_INCLUDE_SYNONYMS, String.valueOf(oracleConfig.isSynonymsEnabled()));
-        connectionProperties.setProperty(OracleConnection.CONNECTION_PROPERTY_REPORT_REMARKS, String.valueOf(oracleConfig.isRemarksReportingEnabled()));
+        @Override
+        protected void configure()
+        {
+            expose(ConnectionFactory.class).annotatedWith(ForBaseJdbc.class);
+            expose(ConnectionFactory.class).annotatedWith(ForRetryingConnectionFactory.class);
+//            install(conditionalModule(
+//                    OracleConfig.class,
+//                    OracleConfig:: isConnectionPoolEnabled,
+//                    innerBinder -> innerBinder.bind(ConnectionFactory.class).annotatedWith(ForBaseJdbc.class).to(OraclePoolConnectionFactory.class).in(SINGLETON),
+//                    innerBinder -> innerBinder.bind(ConnectionFactory.class).annotatedWith(ForBaseJdbc.class).to(RetryingConnectionFactory.class).in(SINGLETON)));
+        }
+
+        @Provides
+        @Singleton
+        @ForBaseJdbc
+        public ConnectionFactory connectionFactory(
+                OracleConfig oracleConfig,
+                OraclePoolConnectionFactory oraclePoolConnectionFactory,
+                RetryingConnectionFactory retryingConnectionFactory)
+        {
+            if (oracleConfig.isConnectionPoolEnabled()) {
+                return oraclePoolConnectionFactory;
+            }
+            return retryingConnectionFactory;
```

Comment:
looks like pre-exisiting logic that we only use `RCF` with non-pooled connections.
I don't know if it was intentional or not, so either we should add `RCF` to pooled connection factory as well or add a comment explaining why we only use `RCF` with non-pooled connections.

Reply:
Sorry, but this ask seems out of scope? This has only been revealed because I slightly refactor the existing code and how it is more evident what the logic is. The scope has been frozen before I started working. If there's a good reason that we should change it, then we will. But we need a good reason.

Acceptance criteria
- no new regressions
- unit tests for new framework that it actually retries
- it is possible to implement retry for any connector in future that fails sometimes to open a connection
- no custom retrying in any jdbc connector, all connectors use new framework

Let me know if I misunderstood AC laid out by @kokosing.